### ISSUE TITLE
Some nodetool commands break when there is an Alternator GSI in the database

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#16153]
We already have an issue about this in a different repository, https://github.com/scylladb/scylla-jmx/issues/226 because the fix is in scylla-jmx and scylla-tools-java, not scylladb core. But I'm nevertheless opening an issue also here because we will need to update the scylla-jmx and scylla-tools-java submodules - and this update will "Fixes" **this** issue. This will allow us to also track the backporting of this update.

Normally, CQL table names may only have alphanumeric characters or underscores. The DynamoDB API also adds dash and dot as allowed characters. But when Alternator creates a GSI or LSI, the materialized view name will also contain a ":" or "!" characters, respectively.

It turns out that JMX, used by nodetool, saves the Java Bean "address" of table names as Java [ObjectName](https://docs.oracle.com/javase/8/docs/api/javax/management/ObjectName.html), which is constructed with a string like `org.apache.cassandra.db:type=tname,keyspace=ksname,columnfamily=cfname`. But when the cfname string contains a colon, it must be quoted. As the ObjectName documentation explains:

>  Each value associated with a key is a string of characters that is either unquoted or quoted.
> An unquoted value is a possibly empty string of characters which may not contain any of the characters comma, equals, colon, or quote.

Of these characters (comma, equals, colon or quote), only the colon is actually possible in table names.

If we don't quote these names, two kinds of bugs can happen:
1. Nodetool may fail when it tries to create an ObjectName to send to the JMX server.
2. The JMX server may fail when it internally tries to list tables and save them as ObjectName.

There are three PRs that fix the known bugs:
* https://github.com/scylladb/scylla-jmx/pull/227
* https://github.com/scylladb/scylla-jmx/pull/230
* https://github.com/scylladb/scylla-tools-java/pull/355

And also a dtest PR that reproduce the bug - and verify the fix - specifically in "nodetool info" and "nodetool tablestates". We may need to write additional tests later (and it's possible we didn't fix all the nodetool bugs in this area...).
* https://github.com/scylladb/scylla-dtest/pull/3738

**Comment by eliransin** [comment]
@scylladb/scylla-maint  can someone handle the backport?
/cc @mykaul

**Comment by denesb** [comment]
> @scylladb/scylla-maint can someone handle the backport? /cc @mykaul

The author is a maintainer himself, so he is the best candidate to do the backport.

**Comment by nyh** [comment]
> > @scylladb/scylla-maint can someone handle the backport? /cc @mykaul
> 
> The author is a maintainer himself, so he is the best candidate to do the backport.

Backporting with all these submodules is no fun, but I accept the challenge. I think because of all these submodules I can't really run CI before merging the various backport, so I hope I don't mess this up.

**Comment by denesb** [comment]
> > > @scylladb/scylla-maint can someone handle the backport? /cc @mykaul
> > 
> > 
> > The author is a maintainer himself, so he is the best candidate to do the backport.
> 
> Backporting with all these submodules is no fun, but I accept the challenge. I think because of all these submodules I can't really run CI before merging the various backport, so I hope I don't mess this up.

If you are not sure, you can create a PR in scylla.git for the submodule update, and merge it yourself after CI passes.
I did the same when I backported the dependency updates.

**Comment by nyh** [comment]
@denesb you're right, but it still requires that I merge everything into the submodules first, without a PR. I'll think what is the least ugly and/or risky way to do this.

**Comment by denesb** [comment]
> @denesb you're right, but it still requires that I merge everything into the submodules first, without a PR. I'll think what is the least ugly and/or risky way to do this.

Yes, you have to cherry-pick the fixes onto the branch-X of the submodule, then push it to origin, and only after that create the PR in sscylla.git. You are right that if CI fails, you already pushed the branches in the submodule. Note that since the submodules are updated with explicit commits, we have a bit more freedom with force-pushes and we can even waive the bisectability requirement for submodule branches. If you find something went wrong, you can backport additional commits, or edit and force-push the existing commit.

**Comment by denesb** [comment]
> > @denesb you're right, but it still requires that I merge everything into the submodules first, without a PR. I'll think what is the least ugly and/or risky way to do this.
> 
> Isn't the core issue that submodules do not have a meaningful CI?

Yes, that is one of the issues. We merge/backport directly to master/branch-X not to next[-X].

**Comment by nyh** [comment]
Merged into master submodules.
PR for the submodule updates:
https://github.com/scylladb/scylladb/pull/16294 (5.4)

**Comment by nyh** [comment]
> Merged into master submodules. PR for the submodule updates: #16294 (5.4) #16296 (5.2)
> 
> When CI passes on those, I can merge them...

The two CI runs passed, and I now merged the tools/java and tools/jmx changes into the scylladb.git next-5.2 and next-5.4 branches. They should hopefully be promoted to branch-5.2 and branch-5.4 soon (which should be successful because the CI already tested that I didn't goof up the backport).

Please note that I did *NOT* backport the dtest commit (see https://github.com/scylladb/scylla-dtest/pull/3738) that verifies that these patches actually fixes the original bug.

**Comment by nyh** [comment]
@kostja what is the meaning of assigning this closed issue to me? As I noted above, I already backported it to 5.2 and 5.4.
What else would you want me to do on this issue?
Did you perhaps intend to assign me some non-open-source-branch issue, not this one?

**Comment by kostja** [comment]
You worked on this issue, so I assigned it to you. It's good for the stats, and for getting notifications (sometimes there are notifications on closed issues).

**Comment by nyh** [comment]
Oh :-)
